### DISCUSSION TITLE
Normalization Layer Support on CoreML EP

### DISCUSSION THREAD

**Discussion by Mertcikla** [microsoft/onnxruntime#21140] [EP Q&A]
I am trying to convert a pytorch model and run inference on an M1 device with C++ using the coreML EP. The model has an InstanceNorm2d layer which the coreML EP doesn't support according to docs and the `onnxruntime.tools.check_onnx_model_mobile_usability` verified that `Unsupported ops: ai.onnx:InstanceNormalization`

I tried converting InstanceNormalization layers to BatchNormalization but the mobile usability outputs  `ai.onnx:15:BatchNormalization`

Here are the logs of attempting to load  
```[V:onnxruntime:, session_state.cc:1152 VerifyEachNodeIsAssignedToAnEp]  Node(s) placed on [CoreMLExecutionProvider]. Number of nodes: 1
[V:onnxruntime:, session_state.cc:1154 VerifyEachNodeIsAssignedToAnEp]   CoreML_13504929322617284055_0 (CoreMLExecutionProvider_CoreML_13504929322617284055_0_0)
[V:onnxruntime:, session_state.cc:1152 VerifyEachNodeIsAssignedToAnEp]  Node(s) placed on [CPUExecutionProvider]. Number of nodes: 1
[V:onnxruntime:, session_state.cc:1154 VerifyEachNodeIsAssignedToAnEp]   InstanceNormalization (/to_gray/to_gray.0/InstanceNormalization)
```

```
[V:onnxruntime:, session_state.cc:1146 VerifyEachNodeIsAssignedToAnEp] Node placements
[V:onnxruntime:, session_state.cc:1152 VerifyEachNodeIsAssignedToAnEp]  Node(s) placed on [CoreMLExecutionProvider]. Number of nodes: 2
[V:onnxruntime:, session_state.cc:1154 VerifyEachNodeIsAssignedToAnEp]   CoreML_17032555989953055218_0 (CoreMLExecutionProvider_CoreML_17032555989953055218_0_0)
[V:onnxruntime:, session_state.cc:1154 VerifyEachNodeIsAssignedToAnEp]   CoreML_17032555989953055218_1 (CoreMLExecutionProvider_CoreML_17032555989953055218_1_1)
[V:onnxruntime:, session_state.cc:1152 VerifyEachNodeIsAssignedToAnEp]  Node(s) placed on [CPUExecutionProvider]. Number of nodes: 1
[V:onnxruntime:, session_state.cc:1154 VerifyEachNodeIsAssignedToAnEp]   BatchNormalization (/to_gray/to_gray.0/BatchNormalization)
```
Here is a small part of the model with the problem, exported to onnx for reproduction [models.zip](https://github.com/user-attachments/files/15932955/reproduce.zip)

Using prebuilt onnxruntime v1.18.0

`python -m onnxruntime.tools.check_onnx_model_mobile_usability model_batchnorm.onnx`

I converted the same model without changes using coremltools and InstanceNorm seem to be supported and verified all layers can execute on M1's NPU.

I don't know how to proceed at this point. Any tips would be appreciated.

**Comment by riccardoalberghi** [comment]
I have the same issue for LayerNorm and GroupNorm. Can something be done?
I also tried to reimplement them from scratch on pytorch, but still no luck in running them on CoreML EP.
